sphinx/domains/python.py

```diff
@@ -285,12 +290,54 @@ def add_target_and_index(self, name_cls, sig, signode):
                                               fullname, '', None))
 
     def before_content(self):
-        # needed for automatic qualification of members (reset in subclasses)
-        self.clsname_set = False
+        # type: () -> None
+        """Handle object nesting before content
+
+        If this class is a nestable object, such as a class object, build up a
+        representation of the nesting heirarchy so that de-nesting multiple
+        levels works correctly.
+
+        If this class isn't a nestable object, just set the current class name
+        using the object prefix, if any. This class name will be removed with
+        :py:meth:`after_content`, and is not added to the list of nested
+        classes.
+        """
+        prefix = None
+        if self.names:
+            (cls_name, cls_name_prefix) = self.names.pop()
+            prefix = cls_name_prefix.strip('.') if cls_name_prefix else None
+            if self.allow_nesting:
+                prefix = cls_name
+        if prefix:
+            self.env.ref_context['py:class'] = prefix
+            if self.allow_nesting:
+                try:
+                    self.env.ref_context['py:classes'].append(prefix)
+                except (AttributeError, KeyError):
+                    self.env.ref_context['py:classes'] = [prefix]
 
     def after_content(self):
-        if self.clsname_set:
-            self.env.ref_context.pop('py:class', None)
+        # type: () -> None
+        """Handle object de-nesting after content
+
+        If this class is a nestable object, removing the last nested class prefix
+        ends further nesting in the object.
+
+        If this class is not a nestable object, the list of classes should not
+        be altered as we didn't affect the nesting levels in
+        :py:meth:`before_content`.
+        """
+        if self.allow_nesting:
+            try:
+                self.env.ref_context['py:classes'].pop()
+            except (KeyError, IndexError):
+                self.env.ref_context['py:classes'] = []
+        try:
+            cls_name = self.env.ref_context.get('py:classes', [])[-1]
+        except IndexError:
+            cls_name = None
+        finally:
+            self.env.ref_context['py:class'] = cls_name
```

Comment:
I think the original is more succinct and pythonic, but I don't have a strong opinion. This change looks fine